Wav2Vec 2.0 Integration

SeanNaren opened the pull request:
Closes #1342. I think the code has been ported over, but I'm sure it can be cleaned up and contains a few bugs.

Haven't tested the wav2vec_asr.py model at all and it's probably broken, but I managed to get pre-training working on AN4, but still need to investigate further!

Any assistance will be appreciated :) cc @VahidooX

VahidooX commented:
I have started the review. It is 2k+ added lines of code and may take some time. In the meanwhile:
+ Please sign off all your commits. Your previous commits should also get signed. The DCO check gives you the commands how to do that.
+Run "python setup.py style --fix" in the main folder to autoformat the style of the code.
+ There are some LGTM warnings to get fixed.
+ You may pull from main to make it up to date as some important changes have happened recently

SeanNaren commented:
It's definitely far from done, and I think I've done a lot of changes here that are not required per say for wav2vec! Hoping to get back to working on this soon

VahidooX commented on nemo/collections/asr/models/wav2vec/modules/gumbel_vector_quantizer.py:
```diff
@@ -0,0 +1,199 @@
+        if not self.time_first:
+            x = x.transpose(1, 2)  # BTC -> BCT
+
+        result["x"] = x
```
Is there a reason result is a dictionary? why not return a list of the outputs form the module?

VahidooX commented on nemo/collections/asr/models/wav2vec/modules/multihead_attention.py:
```diff
@@ -0,0 +1,396 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
NVIDIA licence. Would you please add the link to the main file this code is adopted from?

VahidooX commented on nemo/collections/asr/models/wav2vec/modules/multihead_attention.py:
```diff
@@ -0,0 +1,396 @@
+        return attn, attn_weights
+
+    @staticmethod
+    def _append_prev_key_padding_mask(
```
Are these features used by wav2vec?

VahidooX commented on nemo/collections/asr/models/wav2vec/modules/norm.py:
```diff
@@ -0,0 +1,32 @@
+import torch.nn.functional as F
+
+
+class Fp32GroupNorm(nn.GroupNorm):
```
Why do we need these two modules? Can we call the original modules with such params directly?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_asr_model.py:
```diff
@@ -0,0 +1,51 @@
+from typing import Optional, Union, Dict, List
```
This file looks incomplete for now?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,858 @@
+        return grad * ctx.scale, None
+
+
+class Wav2VecEncoderModel(ModelPT):
```
@okuchaiev which class Wav2VecEncoderModel should inherit from?

okuchaiev replied:
ModelPT seems to be the right choice here because this is something that is trained - so needs to be a LightningModule

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,858 @@
+        return pen
+
+
+class TransposeLast(nn.Module):
```
How about moving modules into the module folder in a file named "wav2vec_modules.py"?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,858 @@
+        return x.transpose(-2, -1)
+
+
+class ConvFeatureExtractionModel(nn.Module):
```
Lets not have model in the name of this class not to confused with the main model. ConvFeatureExtractor?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,858 @@
+        return x
+
+
+class TransformerEncoder(nn.Module):
```
This is not a standard transformerencoder? How about changing the name to something like Wav2vecTransformerEncoder?

VahidooX commented:
> 
> 
> It's definitely far from done, and I think I've done a lot of changes here that are not required per say for wav2vec! Hoping to get back to working on this soon

It is a large PR with more than 2500 lines. It may take some time, and may need more than one round of review. I did a quick review and put some comments. You may take a look at them whenever you can. We can do another round of review when it is closer to be working.

okuchaiev commented on examples/asr/wav2vec_asr.py:
```diff
@@ -0,0 +1,88 @@
+"""
+
+
+@hydra_runner(config_path="conf", config_name="wav2vec_asr")
```
yes, @SeanNaren looks like you forgot to check-in wave2vec_asr.

Also, we should chose a relatively light-weight scenario for pre-training and fine-tuning. 
Should we go with Mozilla's Common Voice for pre-training and then see how it does when fine-tuning using librispeech-clean-100?

SeanNaren commented:
I'm still plugging away at this! Currently working on fixing the _asr part of the model.

SeanNaren commented:
Thanks for all the suggestion @VahidooX @okuchaiev, I addressed most of them but need to follow up on a few, let me give you guys a status update:

Whats been done:

- The code works! I've successfully pre-trained a model on AN4/fine-tuned on AN4, and am working on testing on mozilla common voice right now. Novograd seems to work well, so will be using that instead of adam. I don't have much GPU resources (using AWS atm) but will leave it over the weekend and see how pre-training goes.
- I've cleaned up the code extensively. A lot of the fairseq code seemed very weirdly interconnected so I spent a lot of time ensuring the code flow makes sense, and adding a bit of explanation where necessary. I've removed a lot of modules which were not necessary with latest AMP changes and such, and rely on torch modules where I can.
- Code sits on all existing NeMo code now with no code changes required :)

What needs to be addressed:

- Padding mask is being generated in the forward pass of the wav2vec model. I tested the runtime here and it's very minimal overhead but not vectorized. I'm going to try vectorize it (if someone wants to help, here is the function: https://github.com/SeanNaren/NeMo/blob/feature/wav2vec/nemo/collections/asr/models/wav2vec/wav2vec_model.py#L199-L203)
- I created wav2vec base to wrap basic dataloader functions that I copied from the ctc_model.py. I feel like there is a refactor that can happen to make this an inheritance. I assume NeMo never ran into this because the ctc_model class worked well for all models supported
- As suggested, pull out the fine-tuning portion for a separate PR. But I'll keep it here for the time-being for model training sake as we ensure the codebase works. 

If @VahidooX @okuchaiev can do code reviews again please do!! I'll update with mozilla results as soon as I can.

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,541 @@
+        )  # to NxBxTxC
+        return negs, neg_idxs
+
+    def forward(self, source, padding_mask=None, mask=True, features_only=False) -> Union[tuple, Wav2VecEncoderOutput]:
```
May the values for mask and features_only change for a single model after it is created?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,541 @@
+        self.final_proj = None
+
+
+class ConvFeatureEncoder(nn.Module):
```
You may move ConvFeatureEncoder and Wav2VecTransformerEncoder to the parts folder? @okuchaiev what do you think?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,541 @@
+        padding_mask = self._create_padding_mask(audio_lengths)
+        return self.forward(source, padding_mask, mask=mask, features_only=True)
+
+    def remove_pretraining_modules(self):
```
How about instead of removing those modules, we set a variable to disable them and not erase them. This erasing may cause issues later as we may want to store the final model which do not contain these modules anymore but may be needed to continue the pretraining. When we delete the modules, they are gone.

SeanNaren replied:
Great point :)

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,541 @@
+            groups=conv_cfg.conv_pos_groups,
+        )
+        dropout = 0
+        std = math.sqrt((4 * (1.0 - dropout)) / (conv_cfg.conv_pos * self.embedding_dim))
```
Do these initializations come from the fairseq codebase?

SeanNaren replied:
Yes they do!

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,541 @@
+        if not self.layer_norm_first:
+            x = self.layer_norm(x)
+
+        x = F.dropout(x, p=self.dropout, training=self.training)
```
Is there any reason F.dropout is used instead of nn.Dropout module?

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,541 @@
+                )
+            self.project_inp = nn.Linear(vq_dim, encoder_embed_dim)
+
+        self.mask_emb = nn.Parameter(torch.FloatTensor(encoder_embed_dim).uniform_())
```
Is mask_emb learnable?

SeanNaren replied:
This was ported from the fairseq code, given it's a parameter I'm going to assume so.

SeanNaren commented on examples/asr/experimental/wav2vec/configs/wav2vec_asr.yaml:
```diff
@@ -0,0 +1,71 @@
+model:
+  labels: &labels [ " ", "a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m",
+                    "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "'" ]
+  encoder_path: ???
```
Path to a PT checkpoint, added a comment in the config!

SeanNaren commented on nemo/collections/asr/models/wav2vec/wav2vec_asr_model.py:
```diff
@@ -0,0 +1,174 @@
+        self.vocabulary = cfg.vocabulary
+        self.freeze_encoder_after_steps = cfg.freeze_encoder_after_steps
+        self._num_classes = len(self.vocabulary) + 1
+        self.apply_mask = cfg.mask.apply_mask
```
Would you want me to keep defaults in the config file? I'm happy to add all the param defaults to the config file (taking what's in wav2vec_config.py)

VahidooX replied:
It would be great to have them in the config so user become aware of that and easily update from there.

SeanNaren commented on nemo/collections/asr/models/wav2vec/wav2vec_asr_model.py:
```diff
@@ -0,0 +1,174 @@
+        x = self.final_dropout(x)
+        x = self.proj(x)
+
+        non_padding_mask = ~padding_mask
```
padding_mask is created from the audio_lengths earlier in the calls, it's just a bool tensor representing the masked portions of audio

SeanNaren commented on nemo/collections/asr/models/wav2vec/wav2vec_config.py:
```diff
@@ -0,0 +1,193 @@
+# Copyright (c) 2020, NVIDIA CORPORATION.  All rights reserved.
```
Yep makes sense!

SeanNaren commented:
Great review @VahidooX super thorough which is what I needed. Will take time to address everything but will get there!

SeanNaren commented on nemo/collections/asr/models/wav2vec/wav2vec_asr_model.py:
```diff
@@ -0,0 +1,174 @@
+        super().__init__(cfg=cfg, trainer=trainer)
+
+        schema = OmegaConf.structured(Wav2VecCTCEncoderConfig)
+        cfg = cfg.get('params', {})
```
Yeah I'd like to get rid of this weirdness as well, but how would I be able to with structure configs?

Currently I assume that all parameters associated with `Wav2VecCTCEncoderConfig` are stored within `params` and then merge into the dataclass. 

Was weighing just using the DictConfig out right, and relying on the yaml config. This means worse typing throughout the code though, any thoughts?

SeanNaren commented on examples/asr/experimental/wav2vec/configs/wav2vec.yaml:
```diff
@@ -0,0 +1,66 @@
+    batch_size: 32
+    shuffle: True
+    num_workers: 4
+    labels: [] # No vocab needed for pre-training
```
I'll double check if this is possible, but this was required to use the `audio_to_text_dataset` helper functions, by passing in a config that included the labels. This way I didn't need to duplicate all that logic!

SeanNaren commented on nemo/collections/asr/losses/wav2vecloss.py:
```diff
@@ -0,0 +1,85 @@
+# limitations under the License.
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
```
Let me know what you guys decide on this!

SeanNaren commented on nemo/collections/asr/models/wav2vec/wav2vec_asr_model.py:
```diff
@@ -0,0 +1,174 @@
+        return self._num_classes
+
+
+class Wav2VecASRModel(Wav2VecBase, ASRModel):
```
Yep that would be nice but will require a refactor on top of the current abstractions. I plan on removing the CTC fine-tuning piece from this PR and addressing the ASR part separately, because I think we can do a nicer job there

SeanNaren commented:
Let's try get this over the finish line! @VahidooX I tried my best to answer as much as I can from the review but a lot became outdated. Would you be able to carry out another review on top?

I also tried my best to clear out some of the hydra config stuff and expose them all in the yaml file but you may have a better idea on how to do so, please lemme know!

Finally I'd like to get rid of the `wav2vec_base` class, which is only needed because there is no base class currently that only contains the data loading aspect of `ctc_models.py`. If we make `ctc_models.py` inherit from a similar class, then I think I'll be able to use that base class. Any suggestions?

VahidooX commented on examples/asr/experimental/wav2vec/configs/wav2vec_pretrain.yaml:
```diff
@@ -0,0 +1,122 @@
+      conv_bias: False
+    transformer_encoder:
+      dropout: 0.1
+      conv:
```
Is this conv different from conv_feature_encoder?

SeanNaren replied:
yeah, `conv` is used to generate the positional embeddings for the transformer layers (using a Conv model). The Conv Feature Encoder is used to extract features from raw audio. I added this explanation to the conf :)

VahidooX commented on examples/asr/experimental/wav2vec/configs/wav2vec_pretrain.yaml:
```diff
@@ -0,0 +1,122 @@
+      conv_bias: False
+    transformer_encoder:
+      dropout: 0.1
+      conv:
```
Is it a good idea to drop the conv and encoder levels?

SeanNaren replied:
I think Conv could be dropped, but encoder layers is the most important part imo. Should expose that because I think we'll want a larger model ASAP!

VahidooX commented on examples/asr/experimental/wav2vec/configs/wav2vec_pretrain.yaml:
```diff
@@ -0,0 +1,122 @@
+name: &name "Wav2vec_Pretrain"
+
+model:
+  wav2vec:
```
Is wav2vec level necessary. In other models, we used the model level for the configs of the model. If it is OK, let's make it consistent with others.

VahidooX commented on examples/asr/experimental/wav2vec/configs/wav2vec_pretrain.yaml:
```diff
@@ -0,0 +1,122 @@
+model:
+  wav2vec:
+    loss:
+      prob_ppl_weight: 0.1
```
Do you think it is a good idea to copy the some of the descriptions for each config here so that user may get more familiar with these configs. We can also add some comment on the top of this file and direct them to the config file for more detail. Still maybe not bad to have some comments here. What do you think?

VahidooX commented on examples/asr/experimental/wav2vec/configs/wav2vec_pretrain.yaml:
```diff
@@ -0,0 +1,122 @@
+    loss:
+      prob_ppl_weight: 0.1
+      feature_loss_weight: 0
+    quantize:
```
quantize or quantizer? just a suggestion.

VahidooX commented on examples/asr/experimental/wav2vec/configs/wav2vec_pretrain.yaml:
```diff
@@ -0,0 +1,122 @@
+        dropout: 0.1
+        activation_fn: gelu
+        layer_norm_first: False
+    mask:
```
mask or masking? just a suggestion.

VahidooX commented on nemo/collections/asr/models/wav2vec/wav2vec_model.py:
```diff
@@ -0,0 +1,401 @@
+
+    @typecheck()
+    def forward(self, source, padding_mask=None, mask=True, features_only=False) -> tuple:
+        prob_ppl_loss, cur_codebook_temp = None, None
```
Would you please add have some comments on the the semantics and sizes of the inputs and outputs of the model here?

SeanNaren replied:
I can do so via the `input_types` or `output_types` and add comments there, would that suffice?
